miniapps/electromagnetics/maxwell.cpp

```diff
@@ -0,0 +1,528 @@
+                         abcs, dbcs,
+                         (       dbcs.Size() > 0 ) ? dEdtBCFunc  : NULL
+                        );
+
+   // Display the current number of DoFs in each finite element space
+   Maxwell.PrintSizes();
+
+   // Set the initial conditions for both the electric and magnetic fields
+   VectorFunctionCoefficient EFieldCoef(3,EFieldFunc);
+   VectorFunctionCoefficient BFieldCoef(3,BFieldFunc);
+
+   Maxwell.SetInitialEField(EFieldCoef);
+   Maxwell.SetInitialBField(BFieldCoef);
+
+   // Compute the energy of the initial fields
+   double energy = Maxwell.GetEnergy();
+   if ( mpi.Root() )
+   {
+      cout << "Energy(" << t0 << "ns):  " << energy << "J" << endl;
+   }
+
+   // Approximate the largest stable time step
+   double dtmax = Maxwell.GetMaximumTimeStep();
+
+   // Convert times from nanoseconds to seconds
+   t0 *= tScale_;
+   tf *= tScale_;
+   ts *= tScale_;
+
+   if ( mpi.Root() )
+   {
+      cout << "Maximum Time Step:     " << dtmax / tScale_ << "ns" << endl;
+   }
+
+   // Round down the time step so that tf-t0 is an integer multiple of dt
+   int nsteps = SnapTimeStep(tf-t0, dtsf * dtmax, dt);
+   if ( mpi.Root() )
+   {
+      cout << "Number of Time Steps:  " << nsteps << endl;
+      cout << "Time Step Size:        " << dt / tScale_ << "ns" << endl;
+   }
+
+   // Create the ODE solver
+   SIAVSolver siaSolver(tOrder);
+   siaSolver.Init(Maxwell.GetNegCurl(), Maxwell);
+
+
+   // Initialize GLVis visualization
+   if (visualization)
+   {
+      Maxwell.InitializeGLVis();
+   }
+
+   // Initialize VisIt visualization
+   VisItDataCollection visit_dc("Maxwell-Parallel", &pmesh);
+
+   double t = t0;
+   Maxwell.SetTime(t);
+
+   if ( visit )
+   {
+      Maxwell.RegisterVisItFields(visit_dc);
+   }
+
+   // Write initial fields to disk for VisIt
+   if ( visit )
+   {
+      Maxwell.WriteVisItFields(0);
+   }
+
+   // Send the initial condition by socket to a GLVis server.
+   if (visualization)
+   {
+      Maxwell.DisplayToGLVis();
+   }
+
+   // The main time evolution loop.
+   while (t < tf)
+   {
+      // Run the simulation until a snapshot is needed
+      siaSolver.Run(Maxwell.GetBField(), Maxwell.GetEField(), t, dt, t + ts);
+
+      // Approximate the current energy if the fields
+      energy = Maxwell.GetEnergy();
+      if ( mpi.Root() )
+      {
+         cout << "Energy(" << t/tScale_ << "ns):  " << energy << "J" << endl;
+      }
+
+      // Update local DoFs with current true DoFs
+      Maxwell.SyncGridFuncs();
+
+      // Write fields to disk for VisIt
+      if ( visit )
+      {
+         Maxwell.WriteVisItFields((int)( (t - t0) / ts ) );
+      }
+
+      // Send the solution by socket to a GLVis server.
+      if (visualization)
+      {
+         Maxwell.DisplayToGLVis();
+      }
+   }
+
+   return 0;
+}
+
+// Print the Maxwell ascii logo to the given ostream
+void display_banner(ostream & os)
+{
+   os << "     ___    ____                                      " << endl
```

Comment:
Love the ASCII logo 👍

Reply:
Thanks, I had to make the "M" manually since I wasn't happy with the auto-generated option.